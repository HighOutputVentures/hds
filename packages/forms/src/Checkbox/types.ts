export type Size = "sm" | "md";

export type Variant = "solid" | "outline";
