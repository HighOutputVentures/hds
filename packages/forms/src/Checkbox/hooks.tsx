import { MultiStyleConfig } from "@chakra-ui/react";
import * as React from "react";
import { Size, Variant } from "./types";

type UseCheckboxStyleConfigArg = {
  size: Size;
  variant: Variant;
};

export function useCheckboxStyleConfig(arg: UseCheckboxStyleConfigArg) {
  const styleConfig = React.useMemo<MultiStyleConfig>(
    function getStyleConfig() {
      return {
        parts: ["icon", "container", "control", "label"],
        baseStyle: {
          container: {
            display: "flex",
            alignItems: "start",

            ...(arg.size === "sm" && { gap: "8px" }),
            ...(arg.size === "md" && { gap: "12px" }),
          },
          control: {
            color: "transparent",
            display: "flex",
            alignItems: "center",
            justifyContent: "center",
            flexGrow: 0,
            flexShrink: 0,
            height: "full",
            width: "full",
            border: "1px",
            bgColor: "white",
            borderColor: "neutrals.300",
            transition: "all 300ms ease-in-out",
            _hover: {
              bgColor: "brand.primary.500",
              borderColor: "brand.primary.700",
            },
            _checked: {
              color: "brand.primary.700",
              bgColor: "brand.primary.500",
              borderColor: "brand.primary.700",
            },
            _focus: {},
            _active: {
              boxShadow: "0px 0px 0px 4px #F4EBFF",
            },
            _indeterminate: {
              color: "brand.primary.700",
              bgColor: "brand.primary.500",
              borderColor: "brand.primary.700",
            },
            _disabled: {
              color: "transparent",
              bgColor: "neutrals.100",
              borderColor: "neutrals.200",
              cursor: "not-allowed",
              _active: {
                boxShadow: "none",
              },
              _hover: {},
              _indeterminate: {},
            },

            ...(arg.size === "sm" && {
              width: "16px",
              height: "16px",
              rounded: "4px",
            }),

            ...(arg.size === "md" && {
              width: "20px",
              height: "20px",
              rounded: "6px",
            }),

            ...(arg.variant === "solid" && {
              _checked: {
                color: "white",
                bgColor: "brand.primary.700",
                borderColor: "brand.primary.700",
                _hover: {
                  bgColor: "brand.primary.900",
                  borderColor: "brand.primary.900",
                },
                _disabled: {
                  color: "white",
                  bgColor: "neutrals.300",
                  borderColor: "neutrals.300",
                },
              },

              /* Styles are same as checked */
              _indeterminate: {
                color: "white",
                bgColor: "brand.primary.700",
                borderColor: "brand.primary.700",
                _hover: {
                  bgColor: "brand.primary.900",
                  borderColor: "brand.primary.900",
                },
                _disabled: {
                  color: "white",
                  bgColor: "neutrals.300",
                  borderColor: "neutrals.300",
                },
              },
            }),
          },
          label: {
            margin: "0",
            _disabled: {
              cursor: "not-allowed",
            },
          },
          icon: {
            ...(arg.size === "sm" && {
              width: "10px",
              height: "10px",
            }),

            ...(arg.size === "md" && {
              width: "14px",
              height: "14px",
            }),
          },
        },
        sizes: {},
        variants: {},
        defaultProps: {},
      };
    },
    [arg],
  );

  return styleConfig;
}
