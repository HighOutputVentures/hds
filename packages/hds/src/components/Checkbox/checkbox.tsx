import { MinusIcon } from '@chakra-ui/icons';
import {
  Box,
  Checkbox as ChakraCheckbox,
  CheckboxIcon,
  CheckboxProps,
  Icon,
  IconProps,
  Text,
  useMultiStyleConfig,
  VStack,
} from '@chakra-ui/react';
import React, { ReactNode } from 'react';
import { omit } from './utils';

export type CheckTypeProps = {
  type: 'checkboxType';
  checkbox_icon: 'minusIcon' | 'checkIcon';
};

export type RadioTypeProps = {
  type: 'radioType';
  radio_icon: 'circleIcon' | 'checkIcon';
};

export type CheckProps = (CheckTypeProps | RadioTypeProps) &
  Pick<
    CheckboxProps,
    | 'defaultValue'
    | 'defaultChecked'
    | 'isChecked'
    | 'onChange'
    | 'onFocus'
    | 'isFocusable'
    | 'value'
    | 'isInvalid'
    | 'aria-label'
    | 'size'
  > & {
    helperMsg?: string;
    label?: string | ReactNode;
    onCheck?: (value: boolean) => void;
    variant?: 'primary.outlined' | 'primary.solid';
    disabled?: boolean;
    __testId?: string;
  };

const checkboxHoverStyle = (variant: 'primary.outlined' | 'primary.solid') => {
  return {
    _hover: {
      "span[class*='checkbox__control']:not([data-disabled])": {
        bg: 'brand.primary.500',
        borderColor: 'brand.primary.700',
        color: 'brand.primary.500',
        _checked: {
          bg: `${variant === 'primary.outlined' ? 'brand.primary.500' : 'brand.primary.900'}`,
          borderColor: `${
            variant === 'primary.outlined' ? 'brand.primary.700' : 'brand.primary.900'
          }`,
          color: `${variant === 'primary.outlined' ? 'brand.primary.700' : 'alpha.white.500'}`,
        },
      },
    },
  };
};

const Checkbox = React.forwardRef<HTMLInputElement, CheckProps>((props, ref) => {
  const { size, variant, disabled, ...others } = props;

  const styles = useMultiStyleConfig('Checkbox', { variant, size });

  var classHover = checkboxHoverStyle(variant || 'primary.outlined');

  const checkboxIcon = () => {
    if (props.type === 'checkboxType' && props.checkbox_icon === 'minusIcon') {
      return <MinusIcon data-testid={`hds.checkbox.minus.icon`} />;
    }
    return <CheckboxIcon data-testid={`hds.checkbox.check.icon`} />;
  };

  const radioIcon = () => {
    const iconSize = size === 'md' ? 'xs' : size === 'lg' ? 'md' : size === 'xl' ? '16px' : '8px';
    if (props.type === 'radioType' && props.radio_icon === 'checkIcon') {
      return <Icon as={CheckboxIcon} fontSize={iconSize} />;
    }
    return <CircleIcon fontSize={iconSize} data-testid={`hds.checkbox.circle.icon`} />;
  };

  return (
    <React.Fragment>
      {props.type === 'checkboxType' && (
        <ChakraCheckbox
          variant={variant}
          size={size}
          ref={ref}
          icon={checkboxIcon()}
          isDisabled={disabled}
          sx={classHover}
          data-testid={props.__testId ?? `hds.checkbox`}
          onChange={({ target }) => {
            props.onCheck?.(target.checked);
          }}
          {...omit(others, 'onCheck', 'helperMsg', '__testId')}
        >
          {typeof props.label === 'string' ? (
            <CheckboxLabel checkboxLabel={props.label} helperMsg={props.helperMsg} />
          ) : (
            <Box>{props.label}</Box>
          )}
        </ChakraCheckbox>
      )}
      {props.type === 'radioType' && (
        <ChakraCheckbox
          variant={variant}
          size={size}
          ref={ref}
          data-testid={props.__testId ?? `hds.checkbox`}
          sx={classHover}
          isDisabled={disabled}
          styleConfig={{
            baseStyle: {
              control: {
                ...styles.control,
                borderRadius: 'full',
              },
            },
          }}
          onChange={({ target }) => {
            props.onCheck?.(target.checked);
          }}
          icon={radioIcon()}
          {...omit(others, 'onCheck', 'helperMsg', '__testId')}
        >
          {props.label && typeof props.label === 'string' ? (
            <CheckboxLabel checkboxLabel={props.label} helperMsg={props.helperMsg} />
          ) : (
            props.label && <Box>{props.label}</Box>
          )}
        </ChakraCheckbox>
      )}
    </React.Fragment>
  );
});

export default Checkbox;

interface CheckboxLabelProps {
  checkboxLabel: string;
  helperMsg?: string;
}

const CheckboxLabel = ({ checkboxLabel, helperMsg }: CheckboxLabelProps) => {
  return (
    <VStack align="start" justify="start" spacing={0}>
      <Text size="label-xs-default" color="neutrals.900" data-testid={`hds.checkbox.label`}>
        {checkboxLabel}
      </Text>
      {helperMsg && (
        <Text size="label-xs-default" color="neutrals.600" data-testid={`hds.checkbox.helperMsg`}>
          {helperMsg}
        </Text>
      )}
    </VStack>
  );
};

export const CircleIcon = (props: Omit<IconProps, 'children' | 'css'>) => (
  <Icon
    width="8px"
    height="8px"
    verticalAlign={'center'}
    viewBox="0 0 200 200"
    fill="brand.primary.900"
    {...props}
  >
    <path
      fill="currentColor"
      d="M 100, 100 m -75, 0 a 75,75 0 1,0 150,0 a 75,75 0 1,0 -150,0"
    />
  </Icon>
);
