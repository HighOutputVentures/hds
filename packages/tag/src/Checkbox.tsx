import { Checkbox, CheckboxProps } from '@chakra-ui/react';
import * as React from 'react';

export default React.forwardRef<HTMLInputElement, CheckboxProps>(function HdsCheckbox(props, ref) {
  return <Checkbox ref={ref} aria-label="Select Item" variant="hds" {...props} />;
});
