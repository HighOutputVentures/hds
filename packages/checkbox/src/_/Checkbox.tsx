import {
  Box,
  Checkbox as ChakraCheckbox,
  CheckboxProps as ChakraCheckboxProps,
  Icon,
  Text,
} from "@chakra-ui/react";
import * as React from "react";
import { useCheckboxStyleConfig } from "./hooks";
import CheckIcon from "./icons/CheckIcon";
import MinusIcon from "./icons/MinusIcon";
import { Size, Variant } from "./types";

type BaseProps = Omit<
  ChakraCheckboxProps,
  /* restrict from using these props */
  | "as"
  | "sx"
  | "size"
  | "icon"
  | "variant"
  | "iconSize"
  | "iconColor"
  | "inputProps"
  | "styleConfig"
>;

export interface CheckboxProps extends BaseProps {
  size?: Size;
  hint?: string;
  label?: string;
  variant?: Variant;
  __testId?: string;
}

const Checkbox = React.forwardRef<HTMLInputElement, CheckboxProps>((props, ref) => {
  const {
    size = "md",
    hint,
    label,
    variant = "outline",
    __testId = "hds.checkbox.input",
    ...others
  } = props;

  const styleConfig = useCheckboxStyleConfig({
    size,
    variant,
  });

  const shouldShowLabelSect = !!label || !!hint;

  const CheckboxIcon = React.useMemo(() => {
    if (others.isIndeterminate) {
      return MinusIcon;
    } else {
      return CheckIcon;
    }
  }, []);

  return (
    <ChakraCheckbox
      ref={ref}
      icon={<Icon as={CheckboxIcon} />}
      styleConfig={styleConfig}
      data-testid={__testId}
      {...others}
    >
      {shouldShowLabelSect && (
        <Box>
          <Text
            sx={{
              color: "neutrals.900",
              margin: "0px",
              fontSize: "14px",
              lineHeight: "14px",
              letterSpacing: "0.02em",
            }}
            data-testid="hds.checkbox.label"
          >
            {label}
          </Text>
          <Text
            sx={{
              color: "neutrals.600",
              fontSize: "14px",
              lineHeight: "14px",
              letterSpacing: "0.02em",

              ...(size === "sm" && { marginTop: "0px" }),
              ...(size === "md" && { marginTop: "2px" }),
            }}
            data-testid="hds.checkbox.hint"
          >
            {hint}
          </Text>
        </Box>
      )}
    </ChakraCheckbox>
  );
});

export default Checkbox;
