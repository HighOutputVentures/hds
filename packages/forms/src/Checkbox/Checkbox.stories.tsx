import { Box } from "@chakra-ui/react";
import { ThemeProvider } from "@highoutput/hds/src";
import { ComponentStory, Meta } from "@storybook/react";
import React from "react";
import Checkbox from "./Checkbox";

const Story: Meta<typeof Checkbox> = {
  title: "Components/Checkbox",
  component: Checkbox,
};

export default Story;

const Template: ComponentStory<typeof Checkbox> = (args) => {
  return (
    <ThemeProvider>
      <Box p={4}>
        <Checkbox {...args} />
      </Box>
    </ThemeProvider>
  );
};

export const Default = Template.bind({});

Default.args = {
  ...Default.args,
  size: "md",
  variant: "outline",
  isChecked: false,
  isDisabled: false,
  isIndeterminate: false,
  label: "Remember me",
  hint: "Save my login details for next time.",
};
