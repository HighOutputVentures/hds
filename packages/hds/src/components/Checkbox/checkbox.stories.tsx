import { HStack, VStack } from '@chakra-ui/react';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import React from 'react';

import { ThemeProvider } from '../..';
import Checkbox from './checkbox';

export default {
  title: 'Components/Checkbox',
  component: Checkbox,
  argTypes: {
    disabled: {
      name: 'disabled',
      type: {
        name: 'boolean',
        required: false,
      },
      description: 'true, false',
      defaultValue: false,
      control: 'boolean',
    },
    size: {
      name: 'size',
      type: {
        name: 'string',
        required: false,
      },
      description: `"sm", "md", "lg", "xl"`,
      defaultValue: 'sm',
      control: 'radio',
      options: ['sm', 'md', 'lg', 'xl'],
    },
    type: {
      name: 'checkbox type',
      type: {
        name: 'string',
        required: false,
      },
      defaultValue: 'checkboxType',
      description: 'checkboxType, radioType',
      control: 'radio',
      options: ['checkboxType', 'radioType'],
    },
    checkbox_icon: {
      name: 'checkbox icon type',
      type: {
        name: 'string',
        required: false,
      },
      defaultValue: 'checkIcon',
      control: 'radio',
      if: {
        arg: 'type',
        eq: 'checkboxType',
      },
      options: ['checkIcon', 'minusIcon'],
    },
    radio_icon: {
      name: 'radio icon type',
      type: {
        name: 'string',
        required: false,
      },
      defaultValue: 'circleIcon',
      control: 'radio',
      if: {
        arg: 'type',
        eq: 'radioType',
      },
      options: ['checkIcon', 'circleIcon'],
    },
    variant: {
      name: 'variant',
      type: {
        name: 'string',
        required: false,
      },
      description: "'primary.outlined','primary.solid'",
      defaultValue: 'primary.outlined',
      control: 'radio',
      options: ['primary.outlined', 'primary.solid'],
    },
  },
} as ComponentMeta<typeof Checkbox>;

const Template: ComponentStory<typeof Checkbox> = (args) => {
  return (
    <ThemeProvider>
      <HStack spacing="20px">
        <VStack>
          <VStack mb="20px">
            <Checkbox {...args} />
            <Checkbox {...args} disabled />
            <Checkbox {...args} isChecked />
            <Checkbox {...args} isChecked disabled />
          </VStack>
          <VStack>
            <Checkbox {...args} type="checkboxType" checkbox_icon="minusIcon" />
            <Checkbox
              {...args}
              type="checkboxType"
              checkbox_icon="minusIcon"
              disabled
            />
            <Checkbox
              {...args}
              type="checkboxType"
              checkbox_icon="minusIcon"
              isChecked
            />
            <Checkbox
              {...args}
              type="checkboxType"
              checkbox_icon="minusIcon"
              isChecked
              disabled
            />
          </VStack>
        </VStack>
        <VStack>
          <VStack mb="20px">
            <Checkbox {...args} type="radioType" radio_icon="circleIcon" />
            <Checkbox
              {...args}
              type="radioType"
              radio_icon="circleIcon"
              disabled
            />
            <Checkbox
              {...args}
              type="radioType"
              radio_icon="circleIcon"
              isChecked
            />
            <Checkbox
              {...args}
              type="radioType"
              radio_icon="circleIcon"
              isChecked
              disabled
            />
          </VStack>

          <VStack>
            <Checkbox {...args} type="radioType" radio_icon="checkIcon" />
            <Checkbox
              {...args}
              type="radioType"
              radio_icon="checkIcon"
              disabled
            />
            <Checkbox
              {...args}
              type="radioType"
              radio_icon="checkIcon"
              isChecked
            />
            <Checkbox
              {...args}
              type="radioType"
              radio_icon="checkIcon"
              isChecked
              disabled
            />
          </VStack>
        </VStack>
      </HStack>
    </ThemeProvider>
  );
};

export const Default = Template.bind({});

Default.args = {
  ...Default.args,
  label: 'Remember me',
  helperMsg: 'Save my login details for next time.',
  onCheck: () => {},
};
