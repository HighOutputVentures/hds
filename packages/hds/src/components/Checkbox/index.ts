import Checkbox, { CircleIcon, RadioTypeProps } from '../../components/Checkbox/checkbox';

export { Checkbox, CircleIcon, RadioTypeProps };
