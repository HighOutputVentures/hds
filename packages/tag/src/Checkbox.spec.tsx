import { render } from '@testing-library/react';
import * as React from 'react';
import renderer from 'react-test-renderer';
import Checkbox from './Checkbox';

describe('Checkbox', () => {
  it('Should render checkbox', () => {
    const { queryByRole } = render(<Checkbox />);

    expect(queryByRole('checkbox')).toBeDefined();
  });

  describe('Snapshot', () => {
    it('Should match snapshot', () => {
      const component = renderer.create(<Checkbox />);

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
      component.unmount();
    });
  });
});
