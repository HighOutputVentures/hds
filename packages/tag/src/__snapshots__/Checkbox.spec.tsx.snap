// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Snapshot Should match snapshot 1`] = `
<label
  className="chakra-checkbox css-1elxtiq"
  onClick={[Function]}
>
  <input
    aria-label="Select Item"
    checked={false}
    className="chakra-checkbox__input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    style={
      Object {
        "border": "0",
        "clip": "rect(0, 0, 0, 0)",
        "height": "1px",
        "margin": "-1px",
        "overflow": "hidden",
        "padding": "0",
        "position": "absolute",
        "whiteSpace": "nowrap",
        "width": "1px",
      }
    }
    type="checkbox"
  />
  <span
    aria-hidden={true}
    className="chakra-checkbox__control css-dnty2r"
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
  />
</label>
`;
